### PR TITLE
MUSARK-964 Getting object by UUID endpoint

### PR DESCRIPTION
**Pull request by jarlah**

### CODE REVIEW

File: service_thing_aggregate/app/repositories/dao/ObjectDao.scala [hunk 1]
```diff
@@ -481,6 +481,33 @@ class ObjectDao @Inject()(
...
   }
 
+  def findByUUID(
+    museumId: MuseumId,
+    objectUUID: ObjectUUID,
+    collections: Seq[MuseumCollection]
+  )(implicit currUsr: AuthenticatedUser): Future[MusitResult[Option[MusitObject]]] = {
+    val cids = collections.flatMap(_.schemaIds).distinct
+
+    val q1 = objTable.filter { o =>
```
Comment (ingarabr) [inline service_thing_aggregate/app/repositories/dao/ObjectDao.scala]: By renaming this variable to `queryAllCollecitons` the comment at line 496 might be redundant.

File: service_thing_aggregate/test/controllers/ObjectControllerIntegrationSpec.scala [hunk 2]
```diff
@@ -343,6 +343,22 @@ class ObjectControllerIntegrationSpec extends MusitSpecWithServerPerSuite {
...
+          .withHeaders(fakeToken.asHeader)
+          .withQueryString("collectionIds" -> archeologyCollection)
+          .get()
+          .futureValue
+        res.status mustBe OK
+        val obj = res.json
+        (obj \ "id").as[Int] mustBe 3
+        (obj \ "term").as[String] mustBe "Sommerfugl"
+      }
+    }
```
Comment (kpmeen) [inline service_thing_aggregate/test/controllers/ObjectControllerIntegrationSpec.scala:361]: Missing a test-cases for negative results. What should happen in case nothing is found, or the objectUUID belongs to a different museumId or doesn't exist at all?

File: service_thing_aggregate/test/controllers/ObjectControllerIntegrationSpec.scala [hunk 3]
```diff
@@ -343,6 +343,22 @@ class ObjectControllerIntegrationSpec extends MusitSpecWithServerPerSuite {
         (objects.head \ "subNo").as[String] mustBe "34"
       }
     }
+
+    "finding object by uuid" should {
```
Comment (ingarabr) [inline service_thing_aggregate/test/controllers/ObjectControllerIntegrationSpec.scala]: change spelling, finding -> find

File: service_thing_aggregate/test/repositories/dao/ObjectDaoSpec.scala [hunk 4]
```diff
@@ -522,5 +530,15 @@ class ObjectDaoSpec extends MusitSpecWithAppPerSuite with MusitResultValues {
 
     }
 
+    "finding object by uuid" should {
```
Comment (ingarabr) [inline service_thing_aggregate/test/repositories/dao/ObjectDaoSpec.scala]: finding -> find

File: service_thing_aggregate/test/repositories/dao/ObjectDaoSpec.scala [hunk 5]
```diff
@@ -45,6 +45,14 @@ class ObjectDaoSpec extends MusitSpecWithAppPerSuite with MusitResultValues {
     )
   )
 
+  val lavCollections = Seq(
```
Comment (kpmeen) [inline service_thing_aggregate/test/repositories/dao/ObjectDaoSpec.scala]: translate to English: `lichenCollection`

File: service_thing_aggregate/app/repositories/dao/ObjectDao.scala [hunk 6]
```diff
@@ -481,6 +481,33 @@ class ObjectDao @Inject()(
...
+      o.uuid === objectUUID &&
+      o.museumId === museumId &&
+      o.isDeleted === false
+    }
+    // If use has god mode, look in all collections.
+    val query = if (currUsr.hasGodMode) q1 else q1.filter(_.newCollectionId inSet cids)
+
+    db.run(query.result.headOption)
+      .map {
```
Comment (ingarabr) [inline service_thing_aggregate/app/repositories/dao/ObjectDao.scala]: should we replace {} to () ? It makes the code a bit more compact.

File: service_thing_aggregate/test/repositories/dao/ObjectDaoSpec.scala [hunk 7]
```diff
@@ -522,5 +530,15 @@ class ObjectDaoSpec extends MusitSpecWithAppPerSuite with MusitResultValues {
...
+    "finding object by uuid" should {
+      "return the object" in {
+        val uuid = ObjectUUID.unsafeFromString("dcd37cb7-34ae-484e-a2c0-a1b1925e9b68")
+        val mid = MuseumId(99)
+        val res =
+          dao.findByUUID(mid, uuid, lavCollections).futureValue.successValue.value
+        res.id mustBe ObjectId(51)
+        res.term mustBe "Kartlav"
+      }
+    }
```
Comment (kpmeen) [inline service_thing_aggregate/test/repositories/dao/ObjectDaoSpec.scala]: Missing tests for data not being found.

**Review by kpmeen — CHANGES_REQUESTED** [review]
Some missing test-cases and a minor norglish word

File: service_thing_aggregate/app/repositories/dao/ObjectDao.scala (hunk 6 above)
Comment (kpmeen) [reply service_thing_aggregate/app/repositories/dao/ObjectDao.scala]: yes, and place it one line

File: service_thing_aggregate/test/controllers/ObjectControllerIntegrationSpec.scala (hunk 3 above)
Comment (kpmeen) [reply service_thing_aggregate/test/controllers/ObjectControllerIntegrationSpec.scala]: event better wording `"searching for an object using UUID" should "successfully return the object" in ...`

A good rule of thumb is  that the test wording should read like a proper sentence...that makes sense.

File: service_thing_aggregate/test/controllers/ObjectControllerIntegrationSpec.scala (hunk 2 above)
Comment (jarlah) [inline service_thing_aggregate/test/controllers/ObjectControllerIntegrationSpec.scala]: implemented a test for not found. This test accounts for both not found by UUID, wrong museum or wrong collection

Comment (jarlah) [reply service_thing_aggregate/test/controllers/ObjectControllerIntegrationSpec.scala:361]: done

File: service_thing_aggregate/app/repositories/dao/ObjectDao.scala [hunk 8]
```diff
@@ -481,25 +479,52 @@ class ObjectDao @Inject()(
...
+    val cids = collections.flatMap(_.schemaIds).distinct
+
+    val queryAllCollections = objTable.filter { o =>
+      o.uuid === objectUUID &&
+      o.museumId === museumId &&
+      o.isDeleted === false
+    }
+
+    val query =
+      if (currUsr.hasGodMode) queryAllCollections
```
[the remark this answers is not in the document]
Comment (jarlah) [reply service_thing_aggregate/app/repositories/dao/ObjectDao.scala:496]: not sure if I am going to listen to that

**Review by ingarabr — APPROVED** [review]
